src/uct/sm/scopy/knem/knem_ep.c

```diff
@@ -51,6 +51,43 @@ void uct_knem_iovec_set_buffer(struct knem_cmd_param_iovec *iov, void *buffer)
     iov->base = (uintptr_t)buffer;
 }
 
+static size_t uct_knem_iovec_get_length(struct knem_cmd_param_iovec *iov)
+{
+    return iov->len;
+}
+
+static void* uct_knem_iovec_get_buffer(struct knem_cmd_param_iovec *iov)
+{
+    return (void*)(uintptr_t)iov->base;
+}
+
+static void uct_knem_ep_tx_error(uct_ep_h tl_ep, int knem_fd,
+                                 uct_scopy_tx_op_t tx_op, int ret,
+                                 int icopy_status, int op_errno,
+                                 struct knem_cmd_param_iovec *local_iov,
+                                 size_t local_iov_cnt, uint64_t remote_addr,
+                                 size_t remote_lenth)
+{
+    uct_knem_ep_t *knem_ep = ucs_derived_of(tl_ep, uct_knem_ep_t);
+    UCS_STRING_BUFFER_ONSTACK(knem_fd_strb, 32);
+    UCS_STRING_BUFFER_ONSTACK(ret_strb, 64);
+    struct knem_cmd_param_iovec remote_iov;
+
+    ucs_string_buffer_appendf(&knem_fd_strb, "fd=%d", knem_fd);
+    ucs_string_buffer_appendf(&ret_strb, "ret=%d icopy_status=%d", ret,
+                              icopy_status);
+
+    remote_iov.base = (uintptr_t)remote_addr;
+    remote_iov.len  = remote_lenth;
+
+    uct_scopy_ep_tx_error(
+            &knem_ep->super, &knem_fd_strb, uct_knem_ep_tx_op_str[tx_op],
+            &ret_strb, errno, sizeof(struct knem_cmd_param_iovec),
```

Comment:
pass arg_str and ret_str as const char *

Reply:
done